.werft/platform-delete-preview-environments-cron.ts

```diff
@@ -37,48 +37,102 @@ async function deletePreviewEnvironments() {
-    werft.done("Fetching branches");
-
     werft.phase("Fetching previews");
     let previews: string[]
     try {
         previews = listAllPreviewNamespaces(CORE_DEV_KUBECONFIG_PATH, {});
-        previews.forEach(previewNs => werft.log("Fetching previews", previewNs))
+        previews.forEach(previewNs => werft.log("Fetching preview", previewNs));
+        werft.done("Fetching preview");
     } catch (err) {
-        werft.fail("Fetching previews", err)
+        werft.fail("Fetching preview", err)
     }
-    werft.done("Fetching previews");
+
+    werft.phase("Fetching outdated branches");
+    const branches = getAllBranches();
+    const outdatedPreviews = new Set(branches
+        .filter(branch => {
+          const lastCommit = exec(`git log ${branch} --since=$(date +%Y-%m-%d -d "5 days ago")`, { silent: true })
+          return lastCommit.length < 1
+        })
+        .map(branch => expectedNamespaceFromBranch(branch.replace('origin\/','').trim())))
+
+    const expectedPreviewEnvironmentNamespaces = new Set(branches.map(branch => expectedNamespaceFromBranch(branch)));
 
     werft.phase("deleting previews")
     try {
-        const previewsToDelete = previews.filter(ns => !expectedPreviewEnvironmentNamespaces.has(ns))
-        // Trigger namespace deletion in parallel
-        const promises = previewsToDelete.map(preview => wipePreviewEnvironmentAndNamespace(helmInstallName, preview, CORE_DEV_KUBECONFIG_PATH, { slice: `Deleting preview ${preview}` }));
-        // But wait for all of them to finish before (or one of them to fail) before we continue
-        await Promise.all(promises)
+        const deleteDueToMissingBranch  = previews.filter(ns => !expectedPreviewEnvironmentNamespaces.has(ns))
+        const deleteDueToNoCommitActivity  = previews.filter(ns => outdatedPreviews.has(ns))
+        const deleteDueToNoDBActivity      = previews.filter(ns => checkActivity(ns))
+        const previewsToDelete             = new Set([...deleteDueToMissingBranch, ...deleteDueToNoCommitActivity, ...deleteDueToNoDBActivity])
+
+        if (previewsToDelete.has("staging-main"))
+            previewsToDelete.delete("staging-main")
+
+        const dryRun = true
+        if (dryRun) {
+            previewsToDelete.forEach(preview => werft.log("would have deleted preview environment", preview))
+            werft.done("would have deleted preview environment")
+        }
+        else {
+            previewsToDelete.forEach(preview => werft.log("deleting preview", preview))
+            const promises = previewsToDelete.map(preview => wipePreviewEnvironmentAndNamespace(helmInstallName, preview, CORE_DEV_KUBECONFIG_PATH, { slice: `Deleting preview ${preview}` }));
+            await Promise.all(promises)
+            werft.done("deleting preview")
+        }
     } catch (err) {
-        werft.fail("deleting previews", err)
+        werft.fail("deleting preview", err)
+    }
+}
+
+
+function checkActivity(previewNS) {
```

Comment:
It's preferable to always specify the type for the arguments and the return type of the function, so

```ts
function checkActivity(previewNS: string): boolean {
```

Reply:
Additionally I'd prefer a name that's a bit descriptive of the boolean value that is being returned. Could be something like `isInactive`

Reply:
Finally it would be nice to have a high-level description of what this function checks for. So something like

```
Checks for activity in the preview environments based on data in the underlying database. 
```

Reply:
Done